File: job-queue/internal/retry/retry.go
// Package retry implements retry/backoff decision logic.
//
// Pure decision logic, no I/O: given an attempt count, decide whether a
// failed job should be retried and how long to wait before the next
// attempt. Sleep is injectable so tests never have to wait for real
// backoff durations.
package retry

import (
	"errors"
	"time"
)

// RetryPolicy decides whether to retry a failed job, and how long to
// wait.
type RetryPolicy struct {
	// MaxAttempts is the total number of attempts allowed (including
	// the first, non-retry attempt) before a job is considered
	// exhausted.
	MaxAttempts int
	// BaseDelay is the base delay used by the backoff calculation in
	// NextDelay.
	BaseDelay time.Duration
	// Sleep is an injectable sleep function, defaulting to time.Sleep,
	// so tests can avoid real delays.
	Sleep func(time.Duration)
}

// NewRetryPolicy returns a RetryPolicy with the given maxAttempts and
// baseDelay, and Sleep defaulted to time.Sleep.
func NewRetryPolicy(maxAttempts int, baseDelay time.Duration) *RetryPolicy {
	return &RetryPolicy{
		MaxAttempts: maxAttempts,
		BaseDelay:   baseDelay,
		Sleep:       time.Sleep,
	}
}

// ShouldRetry reports whether a job that just failed on attempt should
// be retried.
//
// TODO(TDD): implement test-first (README.md seam 2).
//
// Expected shape: a job gets p.MaxAttempts total tries, so this is just
// `attempt < p.MaxAttempts`.
func (p *RetryPolicy) ShouldRetry(attempt int) bool {
	panic(errors.New("not implemented"))
}

// NextDelay returns the exponential backoff delay before the next retry
// of attempt.
//
// TODO(TDD): implement test-first (README.md seam 2).
//
// Expected shape: classic exponential backoff — the delay roughly
// doubles with each attempt, e.g. `p.BaseDelay * 2^(attempt-1)`.
// Consider adding jitter (a small random fudge factor) so many
// simultaneously-failing jobs don't all retry in lockstep — not
// required for the first green test, but worth researching.
func (p *RetryPolicy) NextDelay(attempt int) time.Duration {
	panic(errors.New("not implemented"))
}

File: job-queue/src/job_queue/retry.py
"""Retry/backoff decision logic.

Pure decision logic, no I/O: given an attempt count, decide whether a
failed job should be retried and how long to wait before the next
attempt. `clock` and `sleep` are injected so tests never have to wait for
real backoff durations.
"""

import time


class RetryPolicy:
    """Decides whether to retry a failed job, and how long to wait.

    Args:
        max_attempts: total number of attempts allowed (including the
            first, non-retry attempt) before a job is considered
            exhausted.
        base_delay: base delay, in seconds, used by the backoff
            calculation in `next_delay`.
        clock: injectable time source (defaults to `time.monotonic`),
            useful for deterministic tests.
        sleep: injectable sleep function (defaults to `time.sleep`), so
            tests can avoid real delays.
    """

    def __init__(
        self,
        max_attempts: int,
        base_delay: float = 1.0,
        clock=time.monotonic,
        sleep=time.sleep,
    ) -> None:
        self.max_attempts = max_attempts
        self.base_delay = base_delay
        self.clock = clock
        self.sleep = sleep

    def should_retry(self, attempt: int) -> bool:
        """Return True if a job that just failed on `attempt` should be
        retried.

        TODO(TDD): implement test-first (README.md seam 2).
        """
        raise NotImplementedError("implement via TDD")

    def next_delay(self, attempt: int) -> float:
        """Return the exponential backoff delay, in seconds, before the
        next retry of `attempt`.

        TODO(TDD): implement test-first (README.md seam 2).
        """
        raise NotImplementedError("implement via TDD")
